node/actors/bft/src/leader/state_machine.rs

```diff
@@ -93,4 +93,104 @@ impl StateMachine {
         metrics::METRICS.leader_processing_latency[&label].observe_latency(ctx.now() - now);
         Ok(())
     }
+
+    /// In a loop, receives a PrepareQC and sends a LeaderPrepare containing it.
+    /// Every subsequent PrepareQC has to be for a higher view than the previous one (otherwise it
+    /// is skipped). In case payload generation takes too long, some PrepareQC may be elided, so
+    /// that the validator doesn't spend time on generating payloads for already expired views.
+    pub(crate) async fn run_proposer(
+        ctx: &ctx::Ctx,
+        inner: &ConsensusInner,
+        payload_source: &dyn PayloadSource,
+        mut prepare_qc: sync::watch::Receiver<Option<validator::PrepareQC>>,
+    ) -> ctx::Result<()> {
+        let mut next_view = validator::ViewNumber(0);
+        loop {
+            let Some(prepare_qc) = sync::changed(ctx, &mut prepare_qc).await?.clone() else {
+                continue;
+            };
+            if prepare_qc.view() < next_view {
+                continue;
+            };
+            next_view = prepare_qc.view();
+            Self::propose(ctx, inner, payload_source, prepare_qc).await?;
+        }
+    }
+
+    /// Sends a LeaderPrepare for the given PrepareQC.
+    /// Uses `payload_source` to generate a payload if needed.
+    pub(crate) async fn propose(
+        ctx: &ctx::Ctx,
+        inner: &ConsensusInner,
+        payload_source: &dyn PayloadSource,
+        justification: validator::PrepareQC,
+    ) -> ctx::Result<()> {
+        // Get the highest block voted for and check if there's a quorum of votes for it. To have a quorum
+        // in this situation, we require 2*f+1 votes, where f is the maximum number of faulty replicas.
+        let mut count: HashMap<_, usize> = HashMap::new();
+
+        for (vote, signers) in justification.map.iter() {
+            *count.entry(vote.high_vote.proposal).or_default() += signers.len();
+        }
+
+        let highest_vote: Option<validator::BlockHeader> = count
+            .iter()
+            // We only take one value from the iterator because there can only be at most one block with a quorum of 2f+1 votes.
+            .find(|(_, v)| **v > 2 * inner.faulty_replicas())
+            .map(|(h, _)| h)
+            .cloned();
```

Comment:
Nit: This can be shortened to a single `Iterator` combining method:

```rust
.find_map(|(header, v)| {
    (*v > 2 * inner.faulty_replicas()).then_some(*header)
});
```

Reply:
done